base_layer/mmr/src/sparse_merkle_tree/tree.rs

```diff
@@ -253,6 +254,15 @@ impl<H: Digest<OutputSize = U32>> SparseMerkleTree<H> {
         Ok(result)
     }
 
+    /// This will only add new node when it does not exist
+    pub fn insert(&mut self, key: NodeKey, value: ValueHash) -> Result<UpdateResult, SMTError> {
+        if self.get(&key)?.is_some() {
+            return Ok(UpdateResult::AlreadyExistingKey);
```

Comment:
You want this to be an error, I feel.
As written, matching for success becomes more complicated:
```
  match tree.insert(foo) {
    Ok(AlreadyExistingKey) => fail(),
    Ok(Updated(_)) => success(),
    Err(e) => fail(),
}
```

vs

```
  match tree.insert(foo) {
    Ok(_) => successl(),
    Err(_) => fail(),
}
```

Reply:
Indeed, the enum is to differentiate between the different success modes. If you don't care why an (up)sert succeeded you should just be able to check against `.is_ok()`

Reply:
agreed